### PR TITLE
Fix tabbing hidden columns

### LINKED ISSUE #26912: Cannot import exported batch script when sample geometry unticked
**Original reporter:** ISIS/Sarah

### Expected behavior
Exporting a processing table from the GUI should be importable
### Actual behavior
The table cannot be imported as sample geometry is included when the columns are unticked
### Steps to reproduce the behavior
- Export table to batch file
- Check sample geometry was included
- Attempt to import
### Platforms affected
All

### LINKED ISSUE #27028: ISIS SANS Interface Tab key Should Skip Hidden Columns
Currently in the ISIS SANS interface tab moves into hidden columns before moving back to visible ones.

Instead we should skip over any hidden columns

### LINKED ISSUE #27055: Trying to load a run before it exists means it can never be loaded
**Original reporter:** Najet/ISIS

### Expected behavior
If you try to load a run that does not exist yet then it will fail as expected. If you re-try to load it later when it does exist, it should work.

### Actual behavior
Runs are not found the second time around even when they now exist.

### Steps to reproduce the behavior
This needs verifying and will need to be tested at ISIS, but the description of the steps as I understand them are:
- Open the ISIS SANS interface
- Find an upcoming run number where the run hasn't completed yet
- Try to load it and see that it fails
- Wait until the run is complete and check that the file is available on the archive
- Re-try to load it and see that it still fails even though it should be able to find it

### Platforms affected
Windows, possibly all

### LINKED ISSUE #27169: Unhandled exception pressing Shift-Enter
### Expected behavior
Shift-Enter should move to the above cell for editing

### Actual behavior
Shift-Enter sometimes throws an unhandled exception

### Steps to reproduce the behavior
- Open the ISIS SANS interface
- Load the user file MaskFile.txt from the loqdemo example in the ISIS sample data
- Enter a number in the first cell in the table
- Enter a number in the first cell in the second row and press `Shift-Enter`. 
- Mantid often throws an exception here. You may need to repeat the process a few times (change the contents of the second cell and press Shift-Enter).

### Platforms affected
All. Seen in v4.1.0 and master.

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Fixes tabbing between hidden columns, this work includes:
- Moving hidden tabs to the end
- Reworking the SANS code base to not assume positional args
- Introducing a POD class to hold said data
- Updating the CSV parser to not expect an order of data
- Updating various unit tests

**To test:**
- Open the SANS Data GUI with LOQ data and attempt to break it
- For the sample shape issue; add some sample dimensions such as height / width, reload and check they are still there
- Check the various other tabs e.g. beam centre finder

Fixes #27028
Fixes #26912
Fixes #27169 
Fixes #27055 (this is hard to test, but the async race condition no longer exists)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by Mantid-Matthew** [comment]
I see something odd when doing the following:
  * Load in the batch file
  * Process All
  * Show geometry and change the Sample Height
  * Try to reprocess the row

You get an error related to sample shape during the processing.

**Comment by Mantid-Matthew** [comment]
If you click load without a row selected the buttons are disabled and then never re-enabled.

**Comment by Mantid-Matthew** [comment]
If I load the test data and then go to the beam centre tab and click run I get the following error.

Error in execution of algorithm SANSBeamCentreFinder:
to_string
  at line 186 in '/home/matthew/Documents/Mantid/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSBeamCentreFinder.py'
  caused by line 254 in '/home/matthew/Documents/Mantid/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSBeamCentreFinder.py'
  caused by line 390 in '/home/matthew/.local/lib/python2.7/site-packages/enum/__init__.py'
Traceback (most recent call last):
  File "/home/matthew/Documents/Mantid/mantid/scripts/Interface/ui/sans_isis/worker.py", line 39, in run
    self.result = self.func(*self.args, **self.kwargs)
  File "/home/matthew/Documents/Mantid/mantid/scripts/SANS/sans/gui_logic/models/beam_centre_model.py", line 101, in find_beam_centre
    verbose=self.verbose, component=self.component)
  File "/home/matthew/Documents/Mantid/mantid/scripts/SANS/sans/sans_batch.py", line 120, in __call__
    find_direction, verbose, component)
  File "/home/matthew/Documents/Mantid/mantid/scripts/SANS/sans/sans_batch.py", line 127, in _execute_reduction_method
    return centre_finder_new(state, r_min, r_max, max_iter, xstart, ystart, tolerance, find_direction, verbose, component)
  File "/home/matthew/Documents/Mantid/mantid/scripts/SANS/sans/algorithm_detail/centre_finder_new.py", line 76, in centre_finder_new
    beam_centre_alg.execute()
RuntimeError: to_string
  at line 186 in '/home/matthew/Documents/Mantid/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSBeamCentreFinder.py'
  caused by line 254 in '/home/matthew/Documents/Mantid/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSBeamCentreFinder.py'
  caused by line 390 in '/home/matthew/.local/lib/python2.7/site-packages/enum/__init__.py'

File: scripts/Interface/ui/sans_isis/sans_data_processor_gui.py [hunk 1]
```diff
@@ -381,18 +382,20 @@ def _on_wavelength_step_type_changed(self):
...
 
-        self.data_processor_table = jobtreeview.JobTreeView(
-            ["Sample Scatter", "ssp", "Sample Transmission", "stp", "Sample Direct", "sdp",
-             "Can Scatter", "csp",
-             "Can Transmission", "ctp", "Can Direct", "cdp", "Output Name", "User File",
-             "Sample Thickness", "Sample Height", "Sample Width", "Sample Shape",
-             "Options"]
-            , self.cell(""), self)
+        # This dictates the order displayed on the GUI
+        column_names = ["Sample Scatter", "Sample Transmission", "Sample Direct",
```
Comment (Mantid-Matthew) [inline scripts/Interface/ui/sans_isis/sans_data_processor_gui.py]: Could we define this list at the top of the file and lilkewise for period columns e.c.t. so we have all the hardcoded stuff in one place?

**Review by Matthew-Andrew — COMMENTED** [review]
Minor code comments below

**Comment by DavidFair** [comment]
I've opened #27973 for the beam centre issue, as it's a regression in master

**Comment by Mantid-Matthew** [comment]
The sample shape column should be one of three values, Disc, FlatPlate or Cylinder. Currently this is being autoset to an enum value instead and will only accept these enum values.

**Review by Matthew-Andrew — APPROVED** [review]
Looks good to me now

**Pull request merged**